File: native/src/semantic.rs
use crate::parser::{
    BinaryOperator, Expression, Function, Parameter, Program, Statement, Type, UnaryOperator,
};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SemanticError {
    pub message: String,
}

pub fn analyze(program: &Program) -> Result<(), SemanticError> {
    let main = program
        .functions
        .iter()
        .find(|function| function.name == "main")
        .ok_or_else(|| SemanticError {
            message: "program must define 'main'".into(),
        })?;
    for function in &program.functions {
        check_function(function, &program.functions)?;
    }
    check_function(main, &program.functions)
}

fn check_function(function: &Function, functions: &[Function]) -> Result<(), SemanticError> {
    let mut variables = std::collections::HashMap::new();
    for Parameter { name, type_name } in &function.parameters {
        variables.insert(name.clone(), type_name.clone());
    }
    check_block(
        &function.body,
        &mut variables,
        functions,
        function.return_type.clone(),
    )
}

fn check_expression(
    expression: &Expression,
    variables: &std::collections::HashMap<String, Option<Type>>,
    functions: &[Function],
) -> Result<Option<Type>, SemanticError> {
    match expression {
        Expression::String(_) => Ok(Some(Type::String)),
        Expression::Integer(_) => Ok(Some(Type::Int)),
        Expression::Boolean(_) => Ok(Some(Type::Bool)),
        Expression::Array(elements) => {
            let mut inferred: Option<Type> = None;
            for element in elements {
                let element_type = check_expression(element, variables, functions)?;
                if let Some(element_type) = element_type {
                    if let Some(current) = inferred.as_ref() {
                        if *current != element_type {
                            return Err(SemanticError {
                                message: "array elements must share the same type".into(),
                            });
                        }
                    } else {
                        inferred = Some(element_type);
                    }
                }
            }
            Ok(inferred.map(|value| Type::Array(Box::new(value))))
        }
        Expression::Variable(name) => variables.get(name).cloned().ok_or_else(|| SemanticError {
            message: format!("unknown variable '{name}'"),
        }),
        Expression::Binary {
            left,
            operator,
            right,
        } => {
            let left_type = check_expression(left, variables, functions)?;
            let right_type = check_expression(right, variables, functions)?;
            match operator {
                BinaryOperator::Add => {
                    if left_type == right_type && (left_type == Some(Type::Int) || left_type == Some(Type::String)) {
                        Ok(left_type.or(Some(Type::Int)))
                    } else if left_type.is_none() && right_type.is_none() {
                        Ok(Some(Type::Int))
                    } else {
                        Err(SemanticError {
                            message: "+ requires Int or String operands".into(),
                        })
                    }
                }
                BinaryOperator::Subtract
                | BinaryOperator::Multiply
                | BinaryOperator::Divide => {
                    require_types(left_type, right_type, Type::Int, "arithmetic operators")?;
                    Ok(Some(Type::Int))
                }
                BinaryOperator::Greater
                | BinaryOperator::GreaterEqual
                | BinaryOperator::Less
                | BinaryOperator::LessEqual => {
                    require_types(left_type, right_type, Type::Int, "comparison operators")?;
                    Ok(Some(Type::Bool))
                }
                BinaryOperator::Equal | BinaryOperator::NotEqual => {
                    if left_type.is_some() && right_type.is_some() && left_type != right_type {
                        return Err(SemanticError {
                            message: "== requires matching types".into(),
                        });
                    }
                    Ok(Some(Type::Bool))
                }
                BinaryOperator::And | BinaryOperator::Or => {
                    require_types(left_type, right_type, Type::Bool, "logical operators")?;
                    Ok(Some(Type::Bool))
                }
            }
        }
        Expression::Unary {
            operator: UnaryOperator::Not,
            operand,
        } => {
            let operand_type = check_expression(operand, variables, functions)?;
            if operand_type.is_some() && operand_type != Some(Type::Bool) {
                return Err(SemanticError {
                    message: "! requires Bool".into(),
                });
            }
            Ok(Some(Type::Bool))
        }
        Expression::Unary {
            operator: UnaryOperator::Negate,
            operand,
        } => {
            let operand_type = check_expression(operand, variables, functions)?;
            if operand_type.is_some() && operand_type != Some(Type::Int) {
                return Err(SemanticError {
                    message: "unary '-' requires Int".into(),
                });
            }
            Ok(Some(Type::Int))
        }
        Expression::Index { target, index } => {
            let target_type = check_expression(target, variables, functions)?;
            let index_type = check_expression(index, variables, functions)?;
            if index_type.is_some() && index_type != Some(Type::Int) {
                return Err(SemanticError {
                    message: "array index requires Int".into(),
                });
            }
            match target_type {
                Some(Type::Array(inner)) => Ok(Some(*inner)),
                _ => Err(SemanticError {
                    message: "index requires an array".into(),
                }),
            }
        }
        Expression::Call(call) => {
            let function = functions
                .iter()
                .find(|function| function.name == call.name)
                .ok_or_else(|| SemanticError {
                    message: format!("unknown function '{}'", call.name),
                })?;
            if call.arguments.len() != function.parameters.len() {
                return Err(SemanticError {
                    message: format!(
                        "function '{}' expects {} arguments",
                        call.name,
                        function.parameters.len()
                    ),
                });
            }
            for (argument, parameter) in call.arguments.iter().zip(&function.parameters) {
                let argument_type = check_expression(argument, variables, functions)?;
                if parameter.type_name.is_some()
                    && argument_type.is_some()
                    && parameter.type_name != argument_type
                {
                    return Err(SemanticError {
                        message: format!("argument for '{}' has incompatible type", parameter.name),
                    });
                }
            }
            Ok(function.return_type.clone())
        }
    }
}

fn check_block(
    statements: &[Statement],
    variables: &mut std::collections::HashMap<String, Option<Type>>,
    functions: &[Function],
    return_type: Option<Type>,
) -> Result<(), SemanticError> {
    for statement in statements {
        match statement {
            Statement::Let { name, value, .. } => {
                let value_type = check_expression(value, variables, functions)?;
                if let Statement::Let {
                    type_name: Some(declared),
                    ..
                } = statement
                {
                    if value_type.is_some() && value_type != Some(declared.clone()) {
                        return Err(SemanticError {
                            message: format!("variable '{name}' has incompatible type"),
                        });
                    }
                    variables.insert(name.clone(), Some(declared.clone()));
                } else {
                    variables.insert(name.clone(), value_type);
                }
            }
            Statement::Return(value) => {
                let value_type = check_expression(value, variables, functions)?;
                if return_type.is_some() && value_type.is_some() && return_type != value_type {
                    return Err(SemanticError {
                        message: "return value has incompatible type".into(),
                    });
                }
            }
            Statement::If {
                condition,
                then_body,
                else_body,
            } => {
                require_bool(check_expression(condition, variables, functions)?)?;
                check_block(
                    then_body,
                    &mut variables.clone(),
                    functions,
                    return_type.clone(),
                )?;
                check_block(
                    else_body,
                    &mut variables.clone(),
                    functions,
                    return_type.clone(),
                )?;
            }
            Statement::While { condition, body } => {
                require_bool(check_expression(condition, variables, functions)?)?;
                check_block(body, &mut variables.clone(), functions, return_type.clone())?;
            }
            Statement::Assign { target, value } => {
                let target_type = check_expression(target, variables, functions)?;
                let value_type = check_expression(value, variables, functions)?;
                if target_type.is_some() && value_type.is_some() && target_type != value_type {
                    return Err(SemanticError {
                        message: "assignment has incompatible type".into(),
                    });
                }
            }
            Statement::Call(call) if call.name == "print" && call.arguments.len() == 1 => {
                check_expression(&call.arguments[0], variables, functions)?;
            }
            Statement::Call(call) => {
                check_expression(&Expression::Call(call.clone()), variables, functions)?;
            }
        }
    }
    Ok(())
}

fn require_types(
    left: Option<Type>,
    right: Option<Type>,
    expected: Type,
    operation: &str,
) -> Result<(), SemanticError> {
    let expected = Some(expected);
    if (left.is_some() && left != expected) || (right.is_some() && right != expected) {
        return Err(SemanticError {
            message: format!("{operation} require {:?}", expected),
        });
    }
    Ok(())
}

fn require_bool(value: Option<Type>) -> Result<(), SemanticError> {
    if value.is_some() && value != Some(Type::Bool) {
        return Err(SemanticError {
            message: "condition requires Bool".into(),
        });
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::parse;

    #[test]
    fn accepts_hello_program() {
        analyze(&parse("fn main() { print(\"Hello AXIOM\") }").unwrap()).unwrap();
    }

    #[test]
    fn requires_main() {
        let error = analyze(&parse("fn start() { print(\"Hello AXIOM\") }").unwrap()).unwrap_err();
        assert_eq!(error.message, "program must define 'main'");
    }

    #[test]
    fn rejects_mismatched_variable_type() {
        let error = analyze(&crate::parser::parse("fn main() { let value: Int = true }").unwrap())
            .unwrap_err();
        assert!(error.message.contains("incompatible type"));
    }

    #[test]
    fn rejects_non_boolean_condition() {
        let error = analyze(&crate::parser::parse("fn main() { if 1 { print(\"no\") } }").unwrap())
            .unwrap_err();
        assert_eq!(error.message, "condition requires Bool");
    }
}

File: src/axiom/semantic.py
"""Semantic checks for the first AXIOM language slice."""

from __future__ import annotations

from .ast import ArrayLiteral, Assign, Binary, BooleanLiteral, Call, Function, If, Index, IntegerLiteral, Let, Program, Return, StringLiteral, Unary, Variable, While


class SemanticError(ValueError):
    """Raised when a syntactically valid program is not meaningful."""


_BUILTIN_TYPES = {"Int", "Bool", "String"}


def _is_known_type(type_name: str) -> bool:
    return type_name.rstrip("[]") in _BUILTIN_TYPES


def _types_compatible(expected: str, actual: str) -> bool:
    return actual == expected or (actual == "Array" and expected.endswith("[]"))


def analyze(program: Program) -> None:
    function_names: set[str] = set()
    signatures = {}
    for function in program.functions:
        if function.name in function_names:
            raise SemanticError(f"duplicate function '{function.name}'")
        function_names.add(function.name)
        parameter_types = [parameter.type_name for parameter in function.parameters]
        if any(not _is_known_type(type_name) for type_name in parameter_types):
            raise SemanticError(f"function '{function.name}' uses an unknown parameter type")
        if function.return_type is not None and not _is_known_type(function.return_type):
            raise SemanticError(f"function '{function.name}' uses an unknown return type")
        if len({parameter.name for parameter in function.parameters}) != len(function.parameters):
            raise SemanticError(f"function '{function.name}' has duplicate parameters")
        signatures[function.name] = (parameter_types, function.return_type)

    main = next((function for function in program.functions if function.name == "main"), None)
    if main is None:
        raise SemanticError("program must define 'main'")

    for function in program.functions:
        _check_function(function, signatures)


def _check_function(function: Function, signatures) -> None:
    variables = {parameter.name: parameter.type_name for parameter in function.parameters}
    returned = False
    for statement in function.body:
        if isinstance(statement, If):
            condition_type = _expression_type(statement.condition, variables, signatures)
            if condition_type != "Bool":
                raise SemanticError("if condition must be Bool")
            _check_block(statement.then_body, variables.copy(), signatures, function)
            _check_block(statement.else_body, variables.copy(), signatures, function)
            continue
        if isinstance(statement, While):
            if _expression_type(statement.condition, variables, signatures) != "Bool":
                raise SemanticError("while condition must be Bool")
            _check_block(statement.body, variables.copy(), signatures, function)
            continue
        if isinstance(statement, Assign):
            target_type = _expression_type(statement.target, variables, signatures)
            value_type = _expression_type(statement.value, variables, signatures)
            if value_type != target_type:
                raise SemanticError(f"assignment expects {target_type}, got {value_type}")
            continue
        if isinstance(statement, Let):
            value_type = _expression_type(statement.value, variables, signatures)
            if statement.type_name is not None and not _types_compatible(statement.type_name, value_type):
                raise SemanticError(
                    f"variable '{statement.name}' expects {statement.type_name}, got {value_type}"
                )
            if statement.name in variables:
                raise SemanticError(f"variable '{statement.name}' already declared")
            variables[statement.name] = statement.type_name or value_type
            continue
        if not isinstance(statement, Call):
            if isinstance(statement, Return):
                if function.return_type is None:
                    raise SemanticError(f"function '{function.name}' cannot return a value")
                value_type = _expression_type(statement.value, variables, signatures)
                if not _types_compatible(function.return_type, value_type):
                    raise SemanticError(f"function '{function.name}' returns {value_type}, expected {function.return_type}")
                returned = True
                continue
            raise SemanticError(f"unsupported statement in '{function.name}'")
        if statement.name == "print":
            if len(statement.arguments) != 1:
                raise SemanticError("print expects exactly one argument")
            _expression_type(statement.arguments[0], variables, signatures)
            continue
        _check_call(statement, variables, signatures)
    if function.return_type is not None and not returned:
        raise SemanticError(f"function '{function.name}' must return {function.return_type}")


def _check_block(statements, variables, signatures, function):
    for statement in statements:
        if isinstance(statement, Let):
            value_type = _expression_type(statement.value, variables, signatures)
            if statement.type_name is not None and not _types_compatible(statement.type_name, value_type):
                raise SemanticError(f"variable '{statement.name}' expects {statement.type_name}, got {value_type}")
            variables[statement.name] = statement.type_name or value_type
        elif isinstance(statement, Return):
            if function.return_type is None:
                raise SemanticError(f"function '{function.name}' cannot return a value")
            if _expression_type(statement.value, variables, signatures) != function.return_type:
                raise SemanticError(f"function '{function.name}' returns an incompatible value")
        elif isinstance(statement, If):
            if _expression_type(statement.condition, variables, signatures) != "Bool":
                raise SemanticError("if condition must be Bool")
            _check_block(statement.then_body, variables.copy(), signatures, function)
            _check_block(statement.else_body, variables.copy(), signatures, function)
        elif isinstance(statement, While):
            if _expression_type(statement.condition, variables, signatures) != "Bool":
                raise SemanticError("while condition must be Bool")
            _check_block(statement.body, variables.copy(), signatures, function)
        elif isinstance(statement, Assign):
            target_type = _expression_type(statement.target, variables, signatures)
            if _expression_type(statement.value, variables, signatures) != target_type:
                raise SemanticError("assignment has incompatible type")
        elif isinstance(statement, Call):
            if statement.name == "print":
                if len(statement.arguments) != 1:
                    raise SemanticError("print expects exactly one argument")
                _expression_type(statement.arguments[0], variables, signatures)
            else:
                _check_call(statement, variables, signatures)


def _check_call(call: Call, variables: dict[str, str], signatures) -> str:
    if call.name not in signatures:
        raise SemanticError(f"unknown function '{call.name}'")
    parameter_types, return_type = signatures[call.name]
    if len(call.arguments) != len(parameter_types):
        raise SemanticError(f"function '{call.name}' expects {len(parameter_types)} arguments")
    argument_types = [_expression_type(argument, variables, signatures) for argument in call.arguments]
    if argument_types != parameter_types:
        raise SemanticError(f"function '{call.name}' received {argument_types}, expected {parameter_types}")
    if return_type is None:
        raise SemanticError(f"function '{call.name}' has no return value")
    return return_type


def _expression_type(expression, variables: dict[str, str], signatures) -> str:
    if isinstance(expression, StringLiteral):
        return "String"
    if isinstance(expression, IntegerLiteral):
        return "Int"
    if isinstance(expression, BooleanLiteral):
        return "Bool"
    if isinstance(expression, ArrayLiteral):
        if not expression.elements:
            return "Array"
        element_types = [_expression_type(element, variables, signatures) for element in expression.elements]
        if any(element_type != element_types[0] for element_type in element_types[1:]):
            raise SemanticError("array elements must share the same type")
        return f"{element_types[0]}[]"
    if isinstance(expression, Variable):
        if expression.name not in variables:
            raise SemanticError(f"unknown variable '{expression.name}'")
        return variables[expression.name]
    if isinstance(expression, Index):
        target_type = _expression_type(expression.target, variables, signatures)
        if _expression_type(expression.index, variables, signatures) != "Int":
            raise SemanticError("array index requires Int")
        if not target_type.endswith("[]"):
            raise SemanticError("index requires an array")
        return target_type[:-2]
    if isinstance(expression, Call):
        return _check_call(expression, variables, signatures)
    if isinstance(expression, Binary):
        left_type = _expression_type(expression.left, variables, signatures)
        right_type = _expression_type(expression.right, variables, signatures)
        if expression.operator == "+" and left_type == right_type == "Int":
            return "Int"
        if expression.operator in {">", ">=", "<", "<="} and left_type == right_type == "Int":
            return "Bool"
        if expression.operator in {"==", "!="} and left_type == right_type:
            return "Bool"
        if expression.operator in {"-", "*", "/"} and left_type == right_type == "Int":
            return "Int"
        if expression.operator in {"&&", "||"} and left_type == right_type == "Bool":
            return "Bool"
        raise SemanticError(f"operator {expression.operator!r} does not support {left_type} and {right_type}")
    if isinstance(expression, Unary):
        operand_type = _expression_type(expression.operand, variables, signatures)
        if expression.operator == "!" and operand_type == "Bool":
            return "Bool"
        if expression.operator == "-" and operand_type == "Int":
            return "Int"
        raise SemanticError(f"operator {expression.operator!r} does not support {operand_type}")
    raise SemanticError("unsupported expression")